### PR TITLE
validate code on remote terminology service returns nullpointerexception

### PR DESCRIPTION
**Pull request by volodymyr-korzh**
With remote-terminology service enabled, for $validate-code operations, the server responded
with a HTTP 500 and NullPointerException instead of an HTTP 200 and a useful error message if an unknown CodeSystem or unknown ValueSet is provided.
This problem has been fixed.

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/provider/BaseJpaResourceProviderCodeSystem.java [hunk 1]
```diff
@@ -159,9 +160,8 @@ public IBaseParameters validateCode(
 						String code = theCoding.getCode();
 						String display = theCoding.getDisplay();
 
-						result = myValidationSupportChain.validateCode(
-							new ValidationSupportContext(myValidationSupportChain), new ConceptValidationOptions(),
-							codeSystemUrl, code, display, null);
```
Comment (epeartree) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/provider/BaseJpaResourceProviderCodeSystem.java:164]: cool!
but add import ca.uhn.fhir.context.support.IValidationSupport.CodeValidationResult;
so u can do Optional<CodeValidationResult>
and replace 5 instances of IValidationSupport.CodeValidationResult with CodeValidationResult.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/provider/ValueSetOperationProvider.java [hunk 2]
```diff
@@ -138,8 +140,21 @@ public IBaseParameters validateCode(
 				String theDisplayString = (theDisplay != null && theDisplay.hasValue()) ? theDisplay.getValueAsString() : null;
 				String theValueSetUrlString = (theValueSetUrl != null && theValueSetUrl.hasValue()) ?
 					theValueSetUrl.getValueAsString() : null;
-				result = myValidationSupportChain.validateCode(new ValidationSupportContext(myValidationSupportChain),
-					new ConceptValidationOptions(), theSystemString, theCodeString, theDisplayString, theValueSetUrlString);
+				if (theCoding != null) {
+					if (isNotBlank(theCoding.getSystem())) {
+						if (theSystemString != null && !theSystemString.equalsIgnoreCase(theCoding.getSystem())) {
```
Comment (epeartree) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/provider/ValueSetOperationProvider.java:145]: Msg.code nicely added.  thank u for that.

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4RemoteTerminologyTest.java [hunk 3]
```diff
@@ -114,6 +116,45 @@ public void testValidateCodeOperationOnCodeSystem_ByCodingAndUrl_UsingBuiltInCod
 		assertEquals(DISPLAY, respParam.getParameterValue("display").toString());
 	}
 
+	@Test
+	public void testValidateCodeOperationOnCodeSystem_ByCodingAndUrlWhereCodeSystemIsUnknown_ReturnsFalse() {
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4RemoteTerminologyTest.java:120]: is line 121 required? if not please remove
remove line 122

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4RemoteTerminologyTest.java [hunk 4]
```diff
@@ -114,6 +116,45 @@ public void testValidateCodeOperationOnCodeSystem_ByCodingAndUrl_UsingBuiltInCod
...
+		myCodeSystemProvider.myNextReturnCodeSystems = new ArrayList<>();
+		logAllConcepts();
+
+		Parameters respParam = myClient
+			.operation()
+			.onType(CodeSystem.class)
+			.named(JpaConstants.OPERATION_VALIDATE_CODE)
+			.withParameter(Parameters.class, "coding", new Coding()
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4RemoteTerminologyTest.java:128]: the next uri is used in 3 different places.  i'd create a variable at the beginning of the test and use it.  that will greatly improve the test readability.

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4RemoteTerminologyTest.java [hunk 5]
```diff
@@ -114,6 +116,45 @@ public void testValidateCodeOperationOnCodeSystem_ByCodingAndUrl_UsingBuiltInCod
 		assertEquals(DISPLAY, respParam.getParameterValue("display").toString());
 	}
 
+	@Test
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4RemoteTerminologyTest.java:119]: please fix the test name and all the other test name.  
testValidateCodeOperationOnCodeSystem_byCodingAndUrlWhereCodeSystemIsUnknown_returnsFalse

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4RemoteTerminologyTest.java [hunk 6]
```diff
@@ -114,6 +116,45 @@ public void testValidateCodeOperationOnCodeSystem_ByCodingAndUrl_UsingBuiltInCod
...
+
+		String resp = myFhirContext.newXmlParser().setPrettyPrint(true).encodeResourceToString(respParam);
+		ourLog.info(resp);
+
+		assertFalse(((BooleanType) respParam.getParameterValue("result")).booleanValue());
+		assertEquals("Terminology service was unable to provide validation for http://terminology.hl7.org/CodeSystem/uknown-value-set" +
+			"#P", respParam.getParameterValue("message").toString());
+	}
+
+	@Test
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4RemoteTerminologyTest.java:141]: fix test name please.

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4RemoteTerminologyTest.java [hunk 7]
```diff
@@ -163,6 +204,28 @@ public void testValidateCodeOperationOnValueSet_ByUrlSystemAndCode() {
 		assertEquals(DISPLAY_BODY_MASS_INDEX, respParam.getParameterValue("display").toString());
 	}
 
+	@Test
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4RemoteTerminologyTest.java:207]: fix test name

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4RemoteTerminologyTest.java [hunk 8]
```diff
@@ -163,6 +204,28 @@ public void testValidateCodeOperationOnValueSet_ByUrlSystemAndCode() {
...
+	public void testValidateCodeOperationOnValueSet_ByCodingAndUrlWhereValueSetIsUnknown_ReturnsFalse() {
+		myValueSetProvider.myNextReturnValueSets = new ArrayList<>();
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4RemoteTerminologyTest.java:209]: see relevance for line209/210 and remove if need be.

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_8_0/4957-validate-code-on-remote-terminology-returns-null-pointer.yaml [hunk 9]
```diff
@@ -0,0 +1,6 @@
+---
+type: fix
+issue: 4957
```
[the remark this answers is not in the document]
Comment (volodymyr-korzh) [reply hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_8_0/4957-validate-code-on-remote-terminology-returns-null-pointer.yaml:3]: Fixed

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/provider/BaseJpaResourceProviderCodeSystem.java (hunk 1 above)
Comment (volodymyr-korzh) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/provider/BaseJpaResourceProviderCodeSystem.java:164]: Thanks for noticing - fixed

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/provider/ValueSetOperationProvider.java [hunk 10]
```diff
@@ -138,8 +140,21 @@ public IBaseParameters validateCode(
...
+							throw new InvalidRequestException(Msg.code(2352) + "Coding.system '" + theCoding.getSystem() +
+								"' does not equal param system '" + theSystemString + "'. Unable to validate-code.");
+						}
+						theSystemString = theCoding.getSystem();
+						theCodeString = theCoding.getCode();
+						theDisplayString = theCoding.getDisplay();
+					}
+				}
+
```
[the remark this answers is not in the document]
Comment (volodymyr-korzh) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/provider/ValueSetOperationProvider.java:155]: Fixed

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4RemoteTerminologyTest.java (hunk 5 above)
Comment (volodymyr-korzh) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4RemoteTerminologyTest.java:122]: Fixed all test names to comply camelCase convention

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4RemoteTerminologyTest.java (hunk 6 above)
Comment (volodymyr-korzh) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4RemoteTerminologyTest.java:143]: Fixed

**Review by epeartree — APPROVED** [review]

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/provider/ValueSetOperationProvider.java [hunk 11]
```diff
@@ -138,8 +141,21 @@ public IBaseParameters validateCode(
...
+				Optional<CodeValidationResult> resultOptional =
+					validateCodeWithTerminologyService(theSystemString, theCodeString, theDisplayString, theValueSetUrlString);
```
[the remark this answers is not in the document]
Comment (volodymyr-korzh) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/provider/ValueSetOperationProvider.java:157]: Yes, it looks much better this way, but variables `theSystemString, theCodeString, theDisplayString` are not effectively final, so it is not allowed to use them in orElseGet lambda expression

**Pull request merged**